src/storage.rs

```diff
@@ -24,53 +27,282 @@ pub async fn init() -> sqlx::MySqlPool {
-    // async fn mark_messages_recieved(&self, message_id: Vec<u32>);
+    // ) -> Vec<(u32, Vec<u8>)>;
+    // async fn mark_messages_received(&self, message_id: Vec<u32>);
 }
 
 #[cfg(feature = "mysql_db")]
 #[async_trait]
 impl MediatorPersistence for sqlx::MySqlPool {
-    async fn persist_forward_message(&self, forward_msg: &ForwardMsg) {
-        sqlx::query("INSERT INTO forward_raw VALUES (DEFAULT, ?, ?, DEFAULT)")
+    async fn create_account(&self, auth_pubkey: &str) -> Result<(), String> {
+        info!("Adding new account to database with auth_pubkey {:#?}", &auth_pubkey);
+        let insert_res = sqlx::query("INSERT INTO accounts (auth_pubkey) VALUES (?);")
+            .bind(auth_pubkey)
+            .execute(self)
+            .await;
+        if let Err(err) = insert_res {
+            info!("Error during creating new account, {:#?}", err);
+            return Err(format!("{:#}", err))
+        };
+        let account: Vec<u8> = sqlx::query("SELECT (account) FROM accounts WHERE auth_pubkey = ?;")
+            .bind(auth_pubkey)
+            .fetch_one(self)
+            .await
+            .unwrap()
+            .get("account");
+        info!("Created account {:#?}, Adding auth_pubkey {:#?} as a default recipient_key to account", &account, &auth_pubkey);
+        sqlx::query("INSERT INTO recipients (account, recipient_key) VALUES (?, ?);")
+        .bind(&account)
+        .bind(auth_pubkey)
+        .execute(self)
+        .await
+        .unwrap();
+        Ok(())
+    }
+    // async fn vaporize_account(&self, auth_pubkey: String) {
+    //     let account: Vec<u8> = sqlx::query("SELECT (account) FROM accounts WHERE auth_pubkey = ?;")
+    //     .bind(&auth_pubkey)
+    //     .fetch_one(self)
+    //     .await
+    //     .unwrap()
+    //     .get("account");
+
+    //     let mut recipient_rows = sqlx::query(
+    //         "SELECT * FROM recipients WHERE account = ?;"
+    //     )
+    //         .bind(&account)
+    //         .fetch(self);
+
+    //     while let Some(recipient_row) = recipient_rows.try_next().await.unwrap() {
+    //         // map the row into a user-defined domain type
+    //         let recipient: Vec<u8> = recipient_row.get("recipient");  // binary decode
+    //         info!("Recipient {:x?}", recipient);
+    //         sqlx::query("DROP (*) FROM messages WHERE recipient = ?;")
+    //         .bind(&recipient)
+    //         .execute(self)
+    //         .await
+    //         .unwrap();
+    //         sqlx::query("DROP (*) FROM recipients WHERE recipient = ?;")
+    //         .bind(&recipient)
+    //         .execute(self)
+    //         .await
+    //         .unwrap();
+            
+    //     }
+    
+    // }
+    async fn persist_forward_message(&self, forward_msg: &ForwardMsg) -> Result<(), String> {
+        // Fetch recipients with given recipient_key
+        let mut rows = sqlx::query(
+            "SELECT * FROM recipients WHERE recipient_key = ?"
```

Comment:
I also don't see the need for assuming multiple account would register same key. 
44 bytes = 352 bits, thats 9.173994e+105 

If someone is going to DDOS service and try to "register" 1 million of keys per second, it will still take them `3.6717431e+63` ( 2^(32*8)/(1000000 * 60 * 60 * 24 * 365)) years to exhaust the entire space, which is, well damn, less than atoms in universe, but it's a damn long time to be patient for :-D 

We can guard against malicious actors trying to take up key-space without actual possession of private key using different mechanism.

You can make `recipient_key` primary 44 char indexed col and still sleep well at night ;-) Drop `recipient` along with that

Reply:
Had discussed this earlier with Bogdan. There is also a possibility of a mischievous actor (say Bob) registering a known recipient key (say of their conversation partner, Alice). Alice may or may not have pre-registered the key with the mediator before.  I.e. registration of known valid key. 

There would be 4 ways to handle this:

1. Deny registration of the existing recipient_key [_First come, first served_]
2. Allow registration, possibly replacing Alice's registration [_Latest is greatest_]
3. Allow registration, creating duplicate entry  [_Duplicate_]
4. Allow registration of key, scoped to Bob's account. [_Account scoped recipient_keys_]

Consequences:

1. Is simplest, but can lead to 
  - `User enumeration on mediator. Allowing a 3rd party (who obtains the recipient key somehow) to find out if Alice has registered the key on a particular mediator
  - If Alice reuses a private key for multiple parties (which Alice really shouldn't), this is especially more effective.
  - If Alice is reusing keys, then this could also let someone DoS Alice by registering the known key on other mediators where Alice hasn't yet registered. 
2. Is an obvious DoS
3. May result in Bob being able to know if Alice has picked up some message. By asking for status of recipient key. Or ability to mark it as read before Alice can fetch it. 
5. Allows Bob to subscribe to message forwarded to that recipient_key (albeit undecryptable). But avoids DoS, or knowledge if Alice has picked up the message. Because messages will be stored once for each recipient (account scoped recipient_key). Also avoids marking as read, since Alice, Bob will have different copies of the message.

Reply:
Great analysis!

However as per our discussion oncall, we should go with option 1
> Deny registration of the existing recipient_key [First come, first served]

Which does have some corner case risks, but ultimately should be rather be solved on protocol level by requiring proof of key ownership upon key registration.
The issue described are fairly low likely corner cases - typically Alice will register her key in the mediator, before she shares it anywhere else. Alice also definitely should not reuse the same key for different connections. But it's valid corners, however, the protocol level solution should rather address that.

The other 3 options all lead to scenario where Bob can fetch a message address for Alice, be it encrypted, it's classified as personal data - under GDPR and also under other regulations/jurisdictions. So the other 3 options are simply no-go.

Reply:
Note: Syncing of recipient keys would have to be done anyway. So unless `auth_pubkey` is hardware backed and tied to the device, it could possibly also be synced, obviating the need for tying multiple `auth_pubkey` keys to one account.